data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/beacon/GetFinalizedCheckpointState.java

```diff
@@ -0,0 +1,149 @@
+            .response(
+                SC_OK,
+                "Request successful",
+                getResponseType(schemaDefinitionCache),
+                sszResponseType())
+            .withNotFoundResponse()
+            .build());
+    this.chainDataProvider = chainDataProvider;
+  }
+
+  @OpenApi(
+      path = ROUTE,
+      method = HttpMethod.GET,
+      summary = "Get finalized checkpoint state",
+      tags = {TAG_BEACON},
+      description =
+          "Returns full BeaconState object for a finalized checkpoint state from the WS period. \n"
+              + "Depending on `Accept` header it can be returned either as json or as bytes serialized by SSZ.",
+      responses = {
+        @OpenApiResponse(
+            status = RES_OK,
+            content = {
+              @OpenApiContent(type = JSON, from = GetFinalizedCheckpointStateResponse.class),
+              @OpenApiContent(type = OCTET_STREAM)
+            }),
+        @OpenApiResponse(status = RES_BAD_REQUEST),
+        @OpenApiResponse(status = RES_NOT_FOUND),
+        @OpenApiResponse(status = RES_INTERNAL_ERROR),
+        @OpenApiResponse(status = RES_SERVICE_UNAVAILABLE, description = SERVICE_UNAVAILABLE)
+      })
+  @Override
+  public void handle(@NotNull Context ctx) throws Exception {
+    adapt(ctx);
+  }
+
+  @Override
+  public void handleRequest(RestApiRequest request) throws JsonProcessingException {
+    final SafeFuture<Optional<StateAndMetaData>> future =
+        chainDataProvider.getBeaconStateAndMetadata("finalized");
+
+    request.respondAsync(
+        future.thenApply(
+            maybeStateAndMetaData ->
+                maybeStateAndMetaData
+                    .map(
+                        stateAndMetaData -> {
+                          request.header(
+                              HEADER_CONSENSUS_VERSION,
+                              Version.fromMilestone(stateAndMetaData.getMilestone()).name());
+                          return AsyncApiResponse.respondOk(stateAndMetaData);
+                        })
+                    .orElseGet(AsyncApiResponse::respondNotFound)));
+  }
+
+  private static SerializableTypeDefinition<StateAndMetaData> getResponseType(
+      SchemaDefinitionCache schemaDefinitionCache) {
+    return SerializableTypeDefinition.<StateAndMetaData>object()
+        .name("getFinalizedCheckpointStateResponse")
+        .withField("version", MILESTONE_TYPE, ObjectAndMetaData::getMilestone)
+        .withField(EXECUTION_OPTIMISTIC, BOOLEAN_TYPE, ObjectAndMetaData::isExecutionOptimistic)
+        .withField(
+            "data",
+            getSchemaDefinitionForAllMilestones(
+                schemaDefinitionCache,
+                "BeaconState",
+                SchemaDefinitions::getBeaconStateSchema,
+                (beaconState, milestone) ->
+                    schemaDefinitionCache.milestoneAtSlot(beaconState.getSlot()).equals(milestone)),
+            ObjectAndMetaData::getData)
+        .build();
+  }
```

Comment:
i'd be tempted to share the v2 get state here also...

Reply:
Same here, I just removed the `execution_optimistic` field too which makes them slightly different